### PR TITLE
kilkari subscriber with registration report, MA home Page report, cum…

### PR DESCRIPTION
**Pull request by Mandar-Beehyv**
kilkari subscriber with registration date report changes, MA subscriber report  NULL pointer exception , MA home report changes, 
cummulative beneficiary changes.

### CODE REVIEW

**Review by Kantas2601 — APPROVED** [review]

**Review by atishbeehyv123 — CHANGES_REQUESTED** [review]
@Mandar-Beehyv  Can we have sql and hsql as constant which ever part can be isolated without parameters

File: NMSReportingSuite/src/main/java/com/beehyv/nmsreporting/controller/UserController.java [hunk 1]
```diff
@@ -643,26 +643,16 @@ private int countLocationsWithHighCompletion(List<AggregateCumulativeMA> startDa
...
 
-        Long registered = 0L;
-        Long completed =0L;
-
         for (AggregateCumulativeMA end : endData) {
-
             AggregateCumulativeMA start = startMap.get(end.getLocationId());
             if (start != null) {
-                try{
-                    registered = (long) (end.getAshasRegistered() - start.getAshasRegistered());
```
Comment (atishbeehyv123) [inline NMSReportingSuite/src/main/java/com/beehyv/nmsreporting/controller/UserController.java:659]: comments

**Review by atishbeehyv123 — COMMENTED** [review]
```
// Define SQL query segments as constants
private static final String SELECT_SUMS = "SELECT SUM(a.total_subscriptions) AS total_subscriptions, " +
        "SUM(a.subscriptions_active) AS subscriptions_active, " +
        "SUM(a.subscriptions_on_hold) AS subscriptions_on_hold, " +
        "SUM(a.subscriptions_deactivated_LIVE_BIRTH) AS subscriptions_deactivated_LIVE_BIRTH, " +
        "SUM(a.subscriptions_pending) AS subscriptions_pending, " +
        "SUM(a.subscriptions_completed) AS subscriptions_completed, " +
        "SUM(a.subscriptions_rejected) AS subscriptions_rejected, " +
        "SUM(a.Subscriptions_Received_for_PW) AS Subscriptions_Received_for_PW, " +
        "SUM(a.Subscriptions_Received_for_Child) AS Subscriptions_Received_for_Child, " +
        "SUM(a.Subscriptions_Ineligible) AS Subscriptions_Ineligible ";

private static final String FROM_BENEFICIARY = "FROM Beneficiary b INNER JOIN (SELECT beneficiary_id, MAX(subscription_id) AS max_id " +
        "FROM subscriptions s GROUP BY beneficiary_id) max_ids ON b.id = max_ids.beneficiary_id " +
        "INNER JOIN subscriptions s ON max_ids.max_id = s.subscription_id ";

private static final String WHERE_CONDITIONS = "WHERE registration_date >= :fromDate AND registration_date < :toDate AND district_id = :districtId ";

private String buildQueryForBeneficiary() {
    return SELECT_SUMS + FROM_BENEFICIARY + WHERE_CONDITIONS + "GROUP BY health_block_id";
}

private String buildQueryForMotherImportRejection() {
    return SELECT_SUMS + " FROM mother_import_rejection " + WHERE_CONDITIONS + "GROUP BY health_block_id";
}

private String buildQueryForChildImportRejection() {
    return SELECT_SUMS + " FROM child_import_rejection " + WHERE_CONDITIONS + "GROUP BY health_block_id";
}
private List<KilkariSubscriberRegistrationDateDto> executeQuery(String query) {
    List<Object[]> result = entityManager.createNativeQuery(query)
            .setParameter("fromDate", fromDate)
            .setParameter("toDate", toDate)
            .setParameter("districtId", districtId)
            .getResultList();

    return result.stream().map(this::mapRowToDto).collect(Collectors.toList());
}

private KilkariSubscriberRegistrationDateDto mapRowToDto(Object[] row) {
    return new KilkariSubscriberRegistrationDateDto(
            ((BigDecimal) row[0]).intValue(),
            ((BigDecimal) row[1]).intValue(),
            ((BigDecimal) row[2]).intValue(),
            ((BigDecimal) row[3]).intValue(),
            ((BigDecimal) row[4]).intValue(),
            ((BigDecimal) row[5]).intValue(),
            ((BigDecimal) row[6]).intValue(),
            ((BigDecimal) row[7]).intValue(),
            ((BigDecimal) row[8]).intValue(),
            ((BigDecimal) row[9]).intValue(),
            ((BigDecimal) row[10]).intValue(),
            ((BigDecimal) row[11]).intValue()
    );
}
public List<KilkariSubscriberRegistrationDateDto> allCountOffReports(Integer districtId, Date fromDate, Date toDate) {
    List<KilkariSubscriberRegistrationDateDto> beneficiaryData = executeQuery(buildQueryForBeneficiary());
    List<KilkariSubscriberRegistrationDateDto> motherImportData = executeQuery(buildQueryForMotherImportRejection());
    List<KilkariSubscriberRegistrationDateDto> childImportData = executeQuery(buildQueryForChildImportRejection());

    // Combine data as needed or return them separately based on your requirements.
    List<KilkariSubscriberRegistrationDateDto> allData = new ArrayList<>();
    allData.addAll(beneficiaryData);
    allData.addAll(motherImportData);
    allData.addAll(childImportData);
    return allData;
}

```

File: NMSReportingSuite/src/main/java/com/beehyv/nmsreporting/controller/UserController.java (hunk 1 above)
Comment (Mandar-Beehyv) [reply NMSReportingSuite/src/main/java/com/beehyv/nmsreporting/controller/UserController.java:659]: I will add comments and refactoring in another pr.

**Pull request merged**